### ISSUE TITLE
support nanosecond resolution in fluentd

### ISSUE THREAD

**Issue by mr-salty** [fluent/fluentd#461]
Currently fluentd timestamps have second granularity; this is a feature request to support nanoseconds (or milliseconds, although personally I prefer the former).

See also #96 - there was some discussion there of opening a new issue but I don't see one so I thought I would do it.  Proposals there were passing milli/nanoseconds independent of the existing timestamp, or treating the existing timestamp as floating point (currently good enough for 1/4ns granularity assuming unix epoch time).

**Comment by davidwartell** [comment]
+1 for milliseconds in fluentd  I am using Fluentd to store logs in elastic search and events are out of order without better time resolution.  I am now having to shop for an alternative to fluentd

**Comment by repeatedly** [comment]
We have two discussion points.

### millisecond or nanosecond?

From our experience, milliseconds is enough for popular use-cases.
The merit of nanosecond covers almost use-cases.

### How to implement it?

For millisecond, using floating point instead of integer is enough.
For nanosecond, maybe we have 2 ways for serialization format.

- `[sec, usec]` pair

This is straightforward approach.
Ruby's Time has `at(time, usec)` method so easy to handle in fluentd world

- 64bit integer like ktime

This is light-weight approach so serialized data is smaller than above.

In fluentd, we will implement `Fluent::Time` like object to keep backward compatibility.

**Comment by repeatedly** [comment]
I will try to check the performance of nanosecond approach.

**Comment by browny** [comment]
does fluentd support millisecond now?

**Comment by repeatedly** [comment]
Not now. We have a plan to add sub-second support since v0.14.
We are finished to develop v0.12 recently and we are now developing v0.14.

**Comment by repeatedly** [comment]
memo: Need faster TimeParser because cache doesn't work with sub-second support.

**Comment by repeatedly** [comment]
We consider adding nano-second type to msgpack ext in fluentd.
In this way, msgpack chunk handles nano-second time natively.

**Comment by repeatedly** [comment]
@mururu now working on this issue.
If the implementation keeps backward compatibility, we backport the feature to v0.12.

**Comment by repeatedly** [comment]
Continue to discuss this issue via https://github.com/fluent/fluentd/pull/653 PR!

**Comment by mr-salty** [comment]
Hi, do you still plan to backport this to 0.12 per the comment above?

**Comment by whitelynx** [comment]
@mr-salty From how I read his comments, he wasn't planning on backporting; he wanted to wait until the 0.12 release was done before starting work on this. (@repeatedly please correct me if that's wrong)

**Comment by repeatedly** [comment]
@mr-salty Sorry, I missed your comment.
I don't have a backport plan.
If you want to test this feature, please try v0.14.0-pre.1 version.

**Comment by repeatedly** [comment]
Close because this is implemented since v0.14.

**Comment by yissachar** [comment]
@repeatedly Is there anything special that needs to be done to make this work? I can't find any docs covering this.

I'm using v0.14.8 and have the following config:

```
<source>
  @type tail
  read_from_head true
  path /var/log/my.log
  pos_file /var/log/fluentd.pos
  time_format %Y-%m-%d %H:%M:%S.%L
  tag foo.*
</source>

<match foo.**>
  @type aws-elasticsearch-service
  logstash_format true
  flush_interval 5s

  <endpoint>
    url <my-url>
    region <my-region>
  </endpoint>
</match>
```

But I don't end up with millisecond precision - I still only get resolution up to the second. As you can see from my config, I'm using the AWS Elasicsearch plugin, but I don't see why that would make a difference.

**Comment by cosmo0920** [comment]
@yissachar `aws-elasticsearch-service` plugin uses v0.12 API. Using v0.14 API plugin can handle nanosecond precision, but using v0.12 API plugin cannot handle it.
Could you report and request to use v0.14 API at https://github.com/atomita/fluent-plugin-aws-elasticsearch-service repository?

**Comment by yissachar** [comment]
@cosmo0920 Thanks for the tip. I've opened https://github.com/uken/fluent-plugin-elasticsearch/issues/222
